concent_api/core/payments/report.py

```diff
@@ -0,0 +1,202 @@
+    payments: Payments
+
+
+class ClientEthereumAddresses(NamedTuple):
+    provider: str
+    requestor: str
+
+
+class Report(NamedTuple):
+    pair_reports: list
+
+
+def create_report() -> Report:
+    """ Create payments report for Concent. """
+    pair_reports = []
+
+    for subtask in get_subtasks_with_distinct_clients_pairs():  # pylint: disable=not-an-iterable
+        (subtasks_ids, clients_ethereum_addresses) = get_subtask_ids_and_ethereum_addresses_for_clients_pair(
+            subtask.requestor,
+            subtask.provider,
+        )
+        related_pending_responses_for_pair = get_related_pending_responses_for_subtasks_ids(subtasks_ids, subtask.requestor, clients_ethereum_addresses)
+        try:
+            related_payments_for_pair = get_related_payments_for_pair(clients_ethereum_addresses)
+        except SCINotSynchronized:
+            related_payments_for_pair = Payments([], [], [], [], [])
+        is_related_payments_for_pair_empty = related_payments_for_pair.is_empty()  # pylint: disable=no-member
+        is_related_pending_responses_for_pair_empty = related_pending_responses_for_pair.is_empty()  # pylint: disable=no-member
+        if not(is_related_payments_for_pair_empty and is_related_pending_responses_for_pair_empty):  # pylint: disable=no-member
+            pair_report = PairReport(
+                requestor=subtask.requestor,
+                provider=subtask.provider,
+                requestor_ethereum_addresses=[client_ethereum_address.requestor for client_ethereum_address in clients_ethereum_addresses],
+                provider_ethereum_addresses=[client_ethereum_address.provider for client_ethereum_address in clients_ethereum_addresses],
+                pending_responses=related_pending_responses_for_pair,
+                payments=related_payments_for_pair,
+            )
+            pair_reports.append(pair_report)
+
+    return Report(pair_reports)
+
+
+def get_subtasks_with_distinct_clients_pairs() -> QuerySet:
+    """ Get Subtasks with distinct pairs of requestor and provider Clients. """
+    return Subtask.objects.order_by('provider', 'requestor').distinct('provider', 'requestor')
+
+
+def get_subtask_ids_and_ethereum_addresses_for_clients_pair(requestor: Client, provider: Client) -> Tuple:
+    """ Get Subtasks Ids and all Ethereum keys for Clients pair. """
+    subtask_ids = []
+    clients_ethereum_addresses = []
+
+    for subtask in Subtask.objects.filter(requestor=requestor, provider=provider):  # pylint: disable=no-member
+        subtask_ids.append(subtask.subtask_id)
+        task_to_compute = deserialize_message(subtask.task_to_compute.data.tobytes())
+        clients_ethereum_addresses.append(
+            ClientEthereumAddresses(
+                task_to_compute.provider_ethereum_address,
+                task_to_compute.requestor_ethereum_address,
+            )
+        )
+
+    return subtask_ids, list(set(clients_ethereum_addresses))
```

Comment:
If there will be two different `ClientEthereumAddresses` objects but with the same keys will `set()` consider them a duplicate?

Reply:
Yes, if addresses will be duplicated `set()` will do its job :)